src/control/lib/hardware/sysfs/provider.go

```diff
@@ -35,17 +42,95 @@ func (s *Provider) getRoot() string {
 	return s.root
 }
 
+func (s *Provider) sysPath(pathElem ...string) string {
+	pathElem = append([]string{s.getRoot()}, pathElem...)
+
+	return filepath.Join(pathElem...)
+}
+
 // GetNetDevClass fetches the network device class for the given network interface.
 func (s *Provider) GetNetDevClass(dev string) (hardware.NetDevClass, error) {
 	if dev == "" {
 		return 0, errors.New("device name required")
 	}
 
-	devClass, err := ioutil.ReadFile(filepath.Join(s.getRoot(), "class", "net", dev, "type"))
+	devClass, err := ioutil.ReadFile(s.sysPath("class", "net", dev, "type"))
 	if err != nil {
 		return 0, err
 	}
 
 	res, err := strconv.Atoi(strings.TrimSpace(string(devClass)))
 	return hardware.NetDevClass(res), err
 }
+
+// GetTopology builds a minimal topology of network devices from the contents of sysfs.
+func (s *Provider) GetTopology(ctx context.Context) (*hardware.Topology, error) {
```

Comment:
Don't want to over-engineer, but I do think we probably want to use this stuff for more than finding network devices (e.g. storage, etc), so I wonder if maybe the interface should accept a config struct or maybe a bitmask which at a minimum defines which types of devices will be included in the output topology. You could then break the logic below out into smaller helper functions for finding various device types.